### DISCUSSION TITLE
Any way to get better performance (FPS) on i5 10400F cpu with GeForce GT 730 gpu through VMWare Workstation 17?

### DISCUSSION THREAD

**Discussion by DSN9990** [JHRobotics/softgpu#72] [Q&A]
Hi, I'm new here, I'm getting worse performance in VMWare 17 with latest softgpu driver.

I tested 3DMark 2000 (2001 SE didn't work, says DirectX 8 is not installed for some reason), and the performance was worse than in latest VirtualBox, not sure why.

Any help on getting better performance in VMware Workstation 17?

My machine config is here:
[Windows 98.vmx.zip](https://github.com/JHRobotics/softgpu/files/14953047/Windows.98.vmx.zip)

**Comment by DSN9990** [comment]
Any help, please? I will appreciate that.

**Comment by DSN9990** [comment]
Ok, I almost got 3DMark 2001 SE working with [this](https://www.vogons.org/viewtopic.php?p=641489#p641489) patch in order for it to be able to detect DirectX 8, but now it gave me this error:
![Windows 98-2024-04-12-10-51-38](https://github.com/JHRobotics/softgpu/assets/69558812/81936fd8-4c2b-4595-be7e-84562d4c89fb)
Is there any way to work around this? Also, is there any way to get it working without this patch?

**Comment by DSN9990** [comment]
Any help, please? I'm waiting since yesterday.

**Comment by JHRobotics** [comment]
I hear you, just a little patience please! :-)

I see the VM has 1GB of RAM, have you applied this [patch](http://lonecrusader.x10host.com/rloew/patchmem.html)? If this patch is not applied, '98 machines with more than 512 MB of memory do strange things (file system corruption included).

Also, please write which “GL renderer” wgltest/glchecker reports to you (both are on the CD in the tools folder):

![wgltest](https://github.com/JHRobotics/softgpu/assets/105854392/6ba62762-bd1d-4837-9fd1-140afb1e9491)

Also, in 3DMark 2001 currently some tests do not work on VMware (and VirtualBox with vGPU9), I found a bug, but the new release version is currently being tested and if it goes well, I could release a new version in a few hours.

**Reply by DSN9990** [reply]
Thanks! But for some reason, OpenGL is now using llvmpipe, and not SVGA3D. How can I fix that?
This is due to some changes I have made to the C:\WINDOWS\SYSTEM directory, so it is using llvmpipe instead of SVGA3D.
Here is the screenshot:
![Windows 98-2024-04-12-15-50-34](https://github.com/JHRobotics/softgpu/assets/69558812/6afea3e7-2668-4671-b771-41b40d077f43)
And yeah, I have applied patchmem as well, so Win98 works fine now.

**Reply by JHRobotics** [reply]
llvmpipe is used when driver cannot initialize HW 3D subsystem (like when is reported as unsupported), can you please upload here `vmware.log`? (In same folder as virtual machine `*.vmx` file). Maybe here will be logged a reason why 3D is disabled.

**Reply by DSN9990** [reply]
Here is the log file:
[vmware.log](https://github.com/JHRobotics/softgpu/files/14962558/vmware.log)

**Reply by DSN9990** [reply]
It's worth noting that before these changes I made, SVGA3D was working fine (except for performance not reaching 60fps for the most part).

**Reply by DSN9990** [reply]
I forgot to mention that I replaced the original opengl32.dll file from C:\WINDOWS\SYSTEM with the ones provided in the extras folder in the softgpu ISO.

**Reply by JHRobotics** [reply]
Yep, this looks like this is "the problem", because log looks fine. Please don't replace opengl32.dll in SYSTEM folder, some old drivers and wrappers did this, but this creates huge problems with system overal stability. If you don't have original `opengl32.dll`, copy to SYSTEM folder `opengl32.dll` file from SoftGPU CD -> redist/opengl95/DLL - this is older version of MS OpenGL for NT 4.0 and Windows 95, but should works with Windows 98 too.

**Reply by DSN9990** [reply]
Thanks, I got it working now. But my main question here is, how I can get better performance on games such as Quake III, NFS Porsche Unleashed, etc., on 640x480 16 bit mode? I tried these games right now, and the FPS was lower than I expected. I wonder if there are some registry keys to optimize performance or not? And when will be a new release of softgpu?

**Reply by DSN9990** [reply]
And how can I enable dithering for 16 bit color on any Direct3D/OpenGL game? Is that possible or not? If it is possible to do so, they will have a more faithful visual under 16 bit color games, of course.

**Reply by JHRobotics** [reply]
Not much, optimalisation for VMware is this:
```
REGEDIT4

[HKEY_LOCAL_MACHINE\Software\Mesa3D\global]
; VirtualBox doesn't refresh FB with DMA commands, VMware does
"SVGA_DMA_NEED_REREAD"="0"
; VMware cannot blit directly to screen
"SVGA_BLIT_SURF_TO_SCREEN"="0"

[HKEY_LOCAL_MACHINE\Software\VMWSVGA]
; cap RGB565 is real, no need to imaginate some
"RGB565bug"=dword:00000000
```
But it only a bit accelerate 16bit rendering. I'm afraid that I cannot do more here, if you need more performance try VirtualBox - GeForce GT 730 support DX12 and Quake 3 is in 1024x768 on vGPU10 playable.

Currently there is no 16bit dithering, I was thinking of adding something like this, but current rendering is really slow and with this, it’ll be even slower. Also SVGA3D doesn’t have 16-bit rendering support – all operations are done in 32 bit and final render is convert to 16-bit image by CPU – this is reason why most games are in 32bpp faster than in 16bpp although on real HW it is vice versa.

[vmware-optimize.zip](https://github.com/JHRobotics/softgpu/files/14963911/vmware-optimize.zip)

**Reply by JHRobotics** [reply]
New release will be soon (I hope) - today I gradually fount 3 critical bugs in release build, so I’m little tired from it.

**Comment by DSN9990** [comment]
> Not much, optimalisation for VMware is this:
> 
> ```
> REGEDIT4
> 
> [HKEY_LOCAL_MACHINE\Software\Mesa3D\global]
> ; VirtualBox doesn't refresh FB with DMA commands, VMware does
> "SVGA_DMA_NEED_REREAD"="0"
> ; VMware cannot blit directly to screen
> "SVGA_BLIT_SURF_TO_SCREEN"="0"
> 
> [HKEY_LOCAL_MACHINE\Software\VMWSVGA]
> ; cap RGB565 is real, no need to imaginate some
> "RGB565bug"=dword:00000000
> ```
> 
> But it only a bit accelerate 16bit rendering. I'm afraid that I cannot do more here, if you need more performance try VirtualBox - GeForce GT 730 support DX12 and Quake 3 is in 1024x768 on vGPU10 playable.
> 
> Currently there is no 16bit dithering, I was thinking of adding something like this, but current rendering is really slow and with this, it’ll be even slower. Also SVGA3D doesn’t have 16-bit rendering support – all operations are done in 32 bit and final render is convert to 16-bit image by CPU – this is reason why most games are in 32bpp faster than in 16bpp although on real HW it is vice versa.
> 
> [vmware-optimize.zip](https://github.com/JHRobotics/softgpu/files/14963911/vmware-optimize.zip)

Thanks for this information. I will soon upgrade my host to a better GPU with DX12 support, but will this help in any way to get smoother framerate? Do you plan in the future to 
further optimize the performance in VMware too?

**Reply by DSN9990** [reply]
Also why there is no 640x480x32 mode to select in NFS Porsche Unleashed? It only shows 800x600 as the lowest screen mode with 32bit color:
![Windows 98-2024-04-12-19-51-42](https://github.com/JHRobotics/softgpu/assets/69558812/3771fbf4-f1aa-4672-ae82-7ab7b35f7101)

Is there any way to work around this? Maybe through the registry? I do not want to use 800x600x32 because it is slower than 640x480x32, however.

**Reply by DSN9990** [reply]
I was thinking if it is possible to migrate the VMWare VM to VirtualBox to see if there are any performance improvements so far. And which VirtualBox version do you recommend? I tried with the latest version and I got a lot of BSODs/Guru Meditations however.

**Reply by DSN9990** [reply]
OK, I tried migrating to VirtualBox right now and it doesn't seem to boot, I removed a couple of installed hardware through the System control panel, however this did not help. Here is the log and a screenshot:
[VBox.log](https://github.com/JHRobotics/softgpu/files/14964593/VBox.log)
![VirtualBox_Windows 98 SE_12_04_2024_21_07_33](https://github.com/JHRobotics/softgpu/assets/69558812/2e0f43a0-89d7-4508-ae0e-3e3ee934288a)

As you can see, it just gets stuck with a blinking cursor.
Can you help me with this, please? Sorry again for being impatient.
Edit: Here's the config file:
[Windows 98 SE.zip](https://github.com/JHRobotics/softgpu/files/14964685/Windows.98.SE.zip)

**Comment by DSN9990** [comment]
Hi, is there any way to migrate the .vmdk file to VirtualBox like I said above?

**Comment by DSN9990** [comment]
I will close this now due to unrelated answers.
